### PR TITLE
Fix the Issue related to JSON

### LINKED ISSUE #2820: Include JSON as a possibility to be guessed for .txt files
Files sometimes have different content that does not match the file extension.  For example, on Windows depending on the software you have installed, saving a file with JSON content might default to saving it with a `.txt` extension instead of the correct `.json`.  On the internet, files might have various extensions or no extensions when downloaded and saved.  This is not OpenRefine's direct problem, but OpenRefine should be able to inspect (guess) what content a file truly has.

### Proposed solution
`.txt` extension and no extension are 2 of the common ways that files are sometimes found with structured data inside them.
When dealing with a `.txt` file or no extension, it would be preferable for OpenRefine to first read a portion of a file's beginning content and using that as the first contextual clue of what kind of file it is.

For example:
OpenRefine does not guess JSON format when opening a `.txt` file on Windows 10 that is 100% JSON content.  It guesses `line-based` and uses that importer instead, incorrectly.  OpenRefine should instead have guessed JSON as the bestFormat in the rankedFormats, by inspecting the beginning of the file chars are JSON content, maybe 50 chars or more?)

Other file extensions normally don't cause context problems in the wild.  But `.txt` is the most common format and the one that software and some OS's sometimes default to saving files with and as such `.txt` should be treated as that, "a generic bunch of text that first needs to be inspected for some kind of structured data" in order to determine the right kind of importer.

**Describe alternatives you've considered**
Opening the file in Notepad and then looking at the content of the file myself to make a judgement call of what format it might contain and then clicking the appropriate importer selection on OpenRefine's UI.

**Test File**
[Testfile.txt](https://github.com/OpenRefine/OpenRefine/files/4845532/Testfile.txt)

**Additional Context**
Breakpoint on L970 in ImportingUtilities.java seems like a good start for debugging from what I saw and where it gave text/line-based as the best format in the ranked list.

### LINKED ISSUE #2858: XML files not guessed with .txt extension
**Describe the bug**
XML file not be guessed when we change the extension. 

**To Reproduce**
1.  Go to 'OpenRifine'
2. Click on 'Choose files'
4. choose the XML file with .txt extension. [test.txt](https://gist.github.com/djangofan/3186223/raw/ee3840f937e745d5683b046a85cc4a0337e18428/test.xml)

**Current Results**
When I use XML file with .txt Extension :-

Screenshot from 2020-07-01 20-57-16

**Expected behavior**

It will work with or without or any extensions.

**Desktop:**
 - OS: Ubuntu 18.04
 - Browser Version: chromium
 - JRE or JDK Version:  JDK version "11.0.7" 2020-04-14

### PR DESCRIPTION
**Pull request by chetan-v**
Fixes #2820 - Include JSON as a possibility to be guessed for .txt files
Fixes #2858 - XML files not guessed with .txt extension

### CODE REVIEW

**Comment by thadguidry** [comment]
Thanks @chetan-v but this still does not fix the issue in #2820 
I checked out this PR and tried it.
The example provided Test file in #2820 , a `.txt` file, was opened and the OpenRefine data preview did not detect it as JSON content.  Instead, it is still detected as `line-based` content.

image

**Comment by thadguidry** [comment]
@chetan-v Yes, it fixes the JSON parsing for a `.txt` file, but then breaks the parsing for a CSV file that doesn't have an extension.

Download this file that has CSV content inside it.

[TestCSVfile.txt](https://github.com/OpenRefine/OpenRefine/files/4847381/TestCSVfile.txt)

Rename the file to `TestCSVfile` and try again.
Rename the file to `TestCSVfile.csv` and try again.
Rename the file to `TestCSVfile.txt` and try again.

What differences do you observe testing those 3 scenarios?

In all cases the content should be parsed by OpenRefine as **CSV/TSV/separator based file** and should auto-select that importer type instead of **Line-based text files**

**Comment by thadguidry** [comment]
@chetan-v Yes I tested this and seems much improved for detection for CSV and JSON file content, regardless of the filename extension.  Great Job!!! 🥇 

I will open a new issue for what I saw with an XLSX file without a file extension "Colorado-Municipalities-XLSX" that is being detected as Wikitext.

**Comment by chetan-v** [comment]
> @chetan-v Yes I tested this and seems much improved for detection for CSV and JSON file content, regardless of the filename extension. Great Job!!! 
> 
> I will open a new issue for what I saw with an XLSX file without a file extension "Colorado-Municipalities-XLSX" that is being detected as Wikitext.

now what to do test with this xlsv file

**Comment by ostephens** [comment]
Hi @chetan-v - thanks for the contribution. I can see Tom's reviewed the code and I've nothing additional to say there, but just a general comment you've got 6 commits which together result in two lines being changed in a single file. It would be great to tidy this up a bit - you can use the git "rebase" command to do this (this is a good summary https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec)

Often its good to preserve the history of what's happened so I don't habitually use rebase, but in this particular case I think it might make sense

File: main/webapp/modules/core/MOD-INF/controller.js [hunk 1]
```diff
@@ -267,6 +269,7 @@ function registerImporting() {
    *  Mime type to format mappings
    */
   IM.registerMimeType("text/plain", "text/line-based");
```
[the remark this answers is not in the document]
Comment (chetan-v) [reply main/webapp/modules/core/MOD-INF/controller.js:269]: Yes I am also thinking about it.
I am still trying. very soon I get the most probable solution

[the remark this answers is not in the document]
Comment (chetan-v) [reply main/webapp/modules/core/MOD-INF/controller.js]: > Each MIME type can only have a single registration, so this is redundant and will be overwritten by the next line.

I make this change and it work perfectly as I expected but,
XML XLSX file not work  without extension (My changes work on JSON CSV files perfectly)

None of my changes have caused this problem. It was already in openRefine 
Now I am working on that issue.

[the remark this answers is not in the document]
Comment (thadguidry) [reply main/webapp/modules/core/MOD-INF/controller.js]: @chetan-v create a new branch and PR from master for that different XLS/XLSX issue #2850 
I am not sure if we have an issue for the .xml detection without a file extension, if not, you can create one yourself, and then begin work in a new branch and PR.

[the remark this answers is not in the document]
Comment (chetan-v) [reply main/webapp/modules/core/MOD-INF/controller.js]: @thadguidry ok I make new PR but,
please check my new changes.
These are not override anything

[the remark this answers is not in the document]
Comment (thadguidry) [reply main/webapp/modules/core/MOD-INF/controller.js]: Yes I tested the new changes.  Still works correctly.

**Comment by chetan-v** [comment]
These changes also fix #2858 this issue

**Comment by chetan-v** [comment]
@tfmorris please Check I made changes that you suggest.

**Comment by chetan-v** [comment]
@tfmorris ok then I push a fresh Commit on it.

**Comment by chetan-v** [comment]
Now, I make fresh PR

**Pull request closed without merging**